### PR TITLE
feat: voice-chat plugin

### LINKED ISSUE #2375: [Bounty] Clubhouse on top of Twitter
## Catalogue

- Platform
  - [x] Gitcoin
  - [ ] Hackerlink
  - [ ] Mask
- Experience
  - [ ] Beginner
  - [ ] Intermediate
  - [x] Advanced
- Category
  - [ ] Mask Core Feature
  - [x] Mask Plugin
  - [ ] Security/Cryptography Feature
  - [ ] Documentation
  - [ ] Design
  - [ ] Other
    - [ ] /_List it here_/
- Time Commitment
  - [x] 4 weeks
- Bounty Size
  - [x] 2500 USD

## Part 1: Problem Description
Clubhouse has gone viral on the Internet the past week, everywhere in the world. It shows the potential of a socialized audio-only chat room. The success of the Clubhouse is not simply relying on its core audio chat feature. It consists of quite a lot of tweaks and details on their design. Therefore, this bounty is not to copy it. Instead, we just want to show the potential of Mask, being a middle layer on the existing Internet, that can integrate an audio chat room directly on top of Twitter with minimal efforts.

The idea is to inject a chat room in a single tweet, where each person who sees it and have Mask installed can choose to join the room to listen to the discussion, or, raise their hand to speak. In another word, each tweet is a unique chat room, which is designed to be one, with the audio chat functionality. This plugin basically turns each RESTFUL tweet into a REAL TIME discussion. On top of a social media with enough social feature plus the access control system provided by Mask, I believe this would bring our users an alternative choice on the existing social media and it should also bring our (potential) developers more creativeness on how we can play with Mask.

## Part 2: Resources
https://www.joinclubhouse.com/
https://github.com/bestony/neshouse -> an open-source implementation by @bestony in 55 hours

## Part 3: Contact and Support
@yisiliu

### PR DESCRIPTION
**Pull request by venarius**
### Implementation

Instead of using a cloud provider and serverless functions like NESHouse it uses WebRTC and a signaling server. I designed it that way because (imho) it's far less expensive to run because there is no bandwidth usage and the signaling server uses almost no resources because the voice does not go through it but instead is peer-to-peer thanks to WebRTC.

This keeps things very cheap to host and also provides WAY improved security and data protection because the host has no access to any voice parties.

### Screenshots
image
image

Since I mostly only dabble with VueJS feel free to give me some feedback on the code 😅

There is still a small problem I don't know how to fix and for which I need a bit of help: When you open the select element to change your audio input, all the options will be displayed "absolute" on the page. See screenshot:
image

I tried adding the PortalShadowRoot as container but that did not help sadly.

closes #2375

### CODE REVIEW

**Comment by Jack-Works** [comment]
Hello I can help with the shadow root part.

File: packages/voicechat-server/package.json
Comment (Jack-Works) [inline packages/voicechat-server/package.json:2]: Oh wait, so this needs a server to be host on? cc @yisiliu

File: packages/maskbook/src/plugins/Voicechat/types.ts
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Voicechat/types.ts:1]:
```suggestion
export interface VoiceChatMetadata {
```

File: packages/maskbook/src/plugins/Voicechat/types.ts
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Voicechat/types.ts:2]: I suggest not add this kind of property because if the meta exists, the voice chat must be opened. Instead, I suggest allowing custom server URL and chat room name, etc...

File: packages/maskbook/src/plugins/Voicechat/constants.ts
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Voicechat/constants.ts:6]: cc @yisiliu is this stun server ok?

Comment (yisiliu) [reply packages/maskbook/src/plugins/Voicechat/constants.ts:6]: actually I think we can use it for demo but in production, I'd prefer a self-hosted stun.

**Comment by yisiliu** [comment]
I do think this is an alternative approach to building clubhouse on top of twitter and I really like it, since:

1. Voice chat is an extra feature that can be turned on, apart from each tweet itself
2. The topic of each tweet is pre-determined and users can join the chat discussing the content of the Tweet
3. Users would enjoy the freedom of a fully end-to-end voice discussion without worrying about leaving any trace anywhere

Thanks for the great idea and I am truly looking forward to this PR. Thank you @venarius

**Comment by venarius** [comment]
> I do think this is an alternative approach to building clubhouse on top of twitter and I really like it, since:
> 
> 1. Voice chat is an extra feature that can be turned on, apart from each tweet itself
> 2. The topic of each tweet is pre-determined and users can join the chat discussing the content of the Tweet
> 3. Users would enjoy the freedom of a fully end-to-end voice discussion without worrying about leaving any trace anywhere
> 
> Thanks for the great idea and I am truly looking forward to this PR. Thank you @venarius

Thank you! Yeah, that's why I took that route instead of a cloud based one 😃 WebRTC / P2P fits way better to the decentralized and privacy focused approach that crypto currencies offer. And the signaling server / STUN server can easily be hosted on a 5$ Digital-Ocean droplet.

File: packages/maskbook/src/plugins/Voicechat/constants.ts
Comment (venarius) [reply packages/maskbook/src/plugins/Voicechat/constants.ts:6]: https://github.com/coturn/coturn

Comment (venarius) [reply packages/maskbook/src/plugins/Voicechat/constants.ts:6]: You could safely use the google provided one unless you expect a TON of users hitting the voicechat plugin

File: packages/maskbook/src/plugins/Voicechat/types.ts
Comment (venarius) [reply packages/maskbook/src/plugins/Voicechat/types.ts]: Yes, true. I just added an option to provide a custom server URL 🙂

I didn't add option for chat room name because in my opinion it doesn't really make sense to have one when each voice-chat room is already tied to a tweet.

File: packages/voicechat-server/package.json
Comment (venarius) [reply packages/voicechat-server/package.json:2]: Yes, exactly. But this is just a signaling server and no voice is transferred across this server. This has the massive advantage over cloud / SaaS providers in that space in terms of costs.

You can easily deploy this server on a small 5$ Digital-Ocean droplet for example using something like pm2.

File: packages/voicechat-server/tsconfig.json
Comment (Jack-Works) [inline packages/voicechat-server/tsconfig.json:3]: Hi, please cleanup unused comments in this file and mark a comment that this subpackage doesn't have to join the project reference because it is totally isolate.

File: packages/voicechat-server/package.json
Comment (Jack-Works) [inline packages/voicechat-server/package.json:1]: it seems like this package doesn't join the project reference (ts) neither pnpm workspace. I think it's good cause this subpackage doesn't really co-operate with other packages in JS or type level. I guess we should have a new folder for this kind of packages instead of `packages/`. How do you think cc @septs @guanbinrui

File: packages/voicechat-server/package.json
Comment (Jack-Works) [inline packages/voicechat-server/package.json:12]: I guess it should be AGPLv3 or something like that. cc @yisiliu

File: packages/maskbook/src/plugins/Voicechat/types.ts
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Voicechat/types.ts:2]:
```suggestion
    customServer?: string
```

File: packages/maskbook/src/plugins/Voicechat/constants.ts
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Voicechat/constants.ts:2]: It seems like you don't need this line

**Review by Jack-Works — CHANGES_REQUESTED** [review]

File: packages/maskbook/package.json
Comment (venarius) [reply packages/maskbook/package.json]: The socket.io client together with all its dependencies totals to ~90 kb minified. For me it was reasonable enough for all the advantages it provides over normal websocket like automatic reconnects, gzipped data transport, event driven communication, etc...

What do you think?

File: packages/voicechat-server/package.json
Comment (venarius) [reply packages/voicechat-server/package.json:1]: We could create a folder like "ext" or "external" and put it there

**Comment by venarius** [comment]
> @venarius hey man can I have your status upadted please?

Status update on what exactly? Been waiting on further review of my code.

**Comment by yisiliu** [comment]
> > @venarius hey man can I have your status upadted please?
> 
> Status update on what exactly? Been waiting on further review of my code.

oh I am so sorry. can any of you @Jack-Works @septs review the code to push this bounty forward please?

File: packages/maskbook/src/plugins/Voicechat/constants.ts
Comment (septs) [reply packages/maskbook/src/plugins/Voicechat/constants.ts:5]: @yisiliu when deploy my stun server?

File: packages/maskbook/src/plugins/Voicechat/utils/volumeMeter.ts
Comment (septs) [inline packages/maskbook/src/plugins/Voicechat/utils/volumeMeter.ts:9]:
```suggestion
    const fftSize = opts.fftSize ?? 32
    const tweenIn = opts.tweenIn ?? 1.618
    const tweenOut = opts.tweenOut ?? tweenIn * 3
```

File: packages/maskbook/src/plugins/Voicechat/utils/volumeMeter.ts
Comment (septs) [inline packages/maskbook/src/plugins/Voicechat/utils/volumeMeter.ts:11]: why need ignore error?

Comment (septs) [reply packages/maskbook/src/plugins/Voicechat/utils/volumeMeter.ts:11]: plz add note to the line comment

File: packages/voicechat-server/package.json
Comment (septs) [reply packages/voicechat-server/package.json:1]: i think, need be split to https://github.com/DimensionDev/voicechat-server
CC @yisiliu

File: packages/voicechat-server/package.json
Comment (septs) [reply packages/voicechat-server/package.json:2]: i need this server deploy guide.

If you, can provide `Dockerfile` better (We will publish to https://hub.docker.com/u/dimensiondev)

**Review by septs — DISMISSED** [review]

**Comment by venarius** [comment]
@septs I added a dockerfile

File: packages/maskbook/src/plugins/Voicechat/schema.json
Comment (septs) [inline packages/maskbook/src/plugins/Voicechat/schema.json:4]: hmmmmmm, plz add to properties, thanks

File: packages/maskbook/src/plugins/Voicechat/schema.json
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Voicechat/schema.json:10]: Why not number?

File: packages/maskbook/src/plugins/Voicechat/types.ts
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Voicechat/types.ts:2]: this object should match schema.json

File: packages/maskbook/src/plugins/Voicechat/types.ts
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Voicechat/types.ts:9]: What is meter?

File: packages/maskbook/src/plugins/Voicechat/types.ts
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Voicechat/types.ts:8]: can you document those?

File: packages/maskbook/src/plugins/Voicechat/UI/VoicechatDialog.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Voicechat/UI/VoicechatDialog.tsx:48]:
```suggestion
    const [voicechatEnabled, setVoicechatEnabled] = useState(false)
    const [customServerURL, setCustomServerURL] = useState('')
    const [loading, setLoading] = useState(false)
```

**Review by Jack-Works — CHANGES_REQUESTED** [review]

**Comment by septs** [comment]
@venarius https://github.com/DimensionDev/voicechat-server this repo add you as maintain role

plz move `packages/voicechat-server` to the repo.

File: packages/voicechat-server/package.json
Comment (venarius) [reply packages/voicechat-server/package.json:9]: Because not every user will have typescript installed globally. But every user will have typescript inside ```node_modules``` because it's a dependency.

File: packages/voicechat-server/package.json
Comment (venarius) [reply packages/voicechat-server/package.json:18]: Still required now because it's now an external project in another git repo.

File: packages/maskbook/src/plugins/Voicechat/schema.json
Comment (venarius) [reply packages/maskbook/src/plugins/Voicechat/schema.json:10]: Depends: Do you prefer "1.0.0" or "1" ?

**Comment by venarius** [comment]
@septs I moved it

File: packages/voicechat-server/package.json
Comment (Jack-Works) [reply packages/voicechat-server/package.json]: In the scripts field, you can refer to tsc without a global install.

File: packages/maskbook/src/plugins/Voicechat/schema.json
Comment (Jack-Works) [reply packages/maskbook/src/plugins/Voicechat/schema.json:10]: The latter one looks easier to parse

File: packages/maskbook/src/plugins/PluginUI.ts
Comment (Jack-Works) [inline packages/maskbook/src/plugins/PluginUI.ts:32]: Please remove this line

**Comment by Jack-Works** [comment]
@venarius hi i have pushed a commit for stronger typing.

File: packages/maskbook/src/plugins/Voicechat/UI/VoicechatInlay.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Voicechat/UI/VoicechatInlay.tsx:72]: Please do not define a new component inside another component. It will cause the component tree to be totally destroyed and the state is lost.

**Comment by venarius** [comment]
I squashed all commits into one and rebased it on top of master. Seems like this PR doesn't get updated with the changes though.

**Comment by venarius** [comment]
Linting CI fails because of `packages/maskbook/src/extension/options-page/DashboardComponents/WalletAssetsTable.tsx`

File: packages/maskbook/src/plugins/Voicechat/UI/VoicechatDialog.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Voicechat/UI/VoicechatDialog.tsx:63]: this part is a bit strange. do you know you can subscribe to the changes in the `typedMessageMetadata` by

```ts
const currentMeta = useValueRef(getActivatedUI().typedMessageMetadata)
const enabled = currentMeta.keys().includes(...)
```

File: packages/maskbook/src/plugins/Voicechat/UI/VoicechatInlay.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Voicechat/UI/VoicechatInlay.tsx:99]: Hi please do not define nested components. That will cause significant performance problems and real bugs. You may want to learn more details on https://reactjs.org/docs/reconciliation.html#the-diffing-algorithm

**Review by Jack-Works — CHANGES_REQUESTED** [review]

**Comment by Jack-Works** [comment]
by the way, is anyone going to try out the runtime functionality? I'm currently giving reviews only from the code quality perspective.

File: packages/maskbook/src/plugins/PluginUI.ts
Comment (venarius) [reply packages/maskbook/src/plugins/PluginUI.ts:30]: If I do that, I am no longer able to click on the voicechat plugin button when composing a new tweet

Comment (Jack-Works) [reply packages/maskbook/src/plugins/PluginUI.ts:30]: I mean you're adding this plugin twice (also at line 36).

Comment (venarius) [reply packages/maskbook/src/plugins/PluginUI.ts:30]: Is `webOnly` on `Flags` not working correctly in development mode?

Comment (Jack-Works) [reply packages/maskbook/src/plugins/PluginUI.ts:30]: hmm, I don't think so. You can print out to see the value

Comment (venarius) [reply packages/maskbook/src/plugins/PluginUI.ts]: Found it out. Thanks!

**Comment by septs** [comment]
plz rebase master

**Comment by venarius** [comment]
@septs I did a rebase about 2 hours ago. It's already on top of master

**Comment by venarius** [comment]
What's the status here?

cc @septs @Jack-Works @yisiliu

File: packages/maskbook/src/plugins/Voicechat/UI/VoicechatInlay.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Voicechat/UI/VoicechatInlay.tsx:35]: Please give them a more precise type definition, thanks!

File: packages/maskbook/src/plugins/Voicechat/UI/VoicechatInlay.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Voicechat/UI/VoicechatInlay.tsx:40]: Please give them a more precise type definition, thanks!

File: packages/maskbook/src/plugins/Voicechat/UI/VoicechatInlay.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Voicechat/UI/VoicechatInlay.tsx:50]: Please give them a more precise type definition, thanks!

File: packages/maskbook/src/plugins/Voicechat/UI/VoicechatInlay.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Voicechat/UI/VoicechatInlay.tsx:170]: hi, please do not use abbreviations. Use their full versions like "paddingY" or something.

File: packages/maskbook/src/plugins/Voicechat/utils/voicechat.ts
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Voicechat/utils/voicechat.ts:127]: Hi, do you know async-await? It can make then-based code much more clear!

**Review by Jack-Works — CHANGES_REQUESTED** [review]
Hi, I have some style reviews.

File: packages/maskbook/src/plugins/Voicechat/UI/VoicechatDialog.tsx
Comment (Jack-Works) [inline packages/maskbook/src/plugins/Voicechat/UI/VoicechatDialog.tsx:61]: It seems like your code is not updated with the latest master. It is like 

```suggestion
    const currentMeta = useValueRef(globalTypedMessageMetadata)
```

in the new flavor

Comment (Jack-Works) [reply packages/maskbook/src/plugins/Voicechat/UI/VoicechatDialog.tsx:61]: And `getActivatedUI()` changed to `activatedSocialNetworkUI`

**Comment by septs** [comment]
@venarius Voice Chat Server-side business logic stable now?

**Comment by yisiliu** [comment]
And if so, we are going to deploy the server and merge this into master by the end of this week. Sorry about the delay. This feature has been stalled by some other features in the dashboard and we will pay out the bounty as soon as we have tested it out. Thanks for your contribution!

**Comment by venarius** [comment]
> @venarius Voice Chat Server-side business logic stable now?

Yes

**Comment by septs** [comment]
plz fix all error and warning.

**Comment by venarius** [comment]
> plz fix all error and warning. (ready to merge?)

done 👍

**Comment by yisiliu** [comment]
@venarius thanks! this bounty is paid out and merged soon.

**Comment by septs** [comment]
@venarius plz rebase to `develop` branch, and migrate new plugin infra (see #3066).

sorry, we rebase after merge the pr.

**Comment by yisiliu** [comment]
@septs no you should do the rebase instead the pr creator.

**Comment by septs** [comment]
Replaced by #3518

**Pull request closed without merging**
